File: my_mindhelp_app/lib/models/quiz.dart
import 'package:json_annotation/json_annotation.dart';

part 'quiz.g.dart';

@JsonSerializable()
class Quiz {
  final String id;
  final String title;
  final String description;
  final List<QuizQuestion> questions;
  final String category;

  Quiz({
    required this.id,
    required this.title,
    required this.description,
    required this.questions,
    required this.category,
  });

  factory Quiz.fromJson(Map<String, dynamic> json) => _$QuizFromJson(json);
  Map<String, dynamic> toJson() => _$QuizToJson(this);
}

@JsonSerializable()
class QuizQuestion {
  final String id;
  final String question;
  final List<QuizOption> options;
  final int order;

  QuizQuestion({
    required this.id,
    required this.question,
    required this.options,
    required this.order,
  });

  factory QuizQuestion.fromJson(Map<String, dynamic> json) => _$QuizQuestionFromJson(json);
  Map<String, dynamic> toJson() => _$QuizQuestionToJson(this);
}

@JsonSerializable()
class QuizOption {
  final String id;
  final String text;
  final int score;

  QuizOption({
    required this.id,
    required this.text,
    required this.score,
  });

  factory QuizOption.fromJson(Map<String, dynamic> json) => _$QuizOptionFromJson(json);
  Map<String, dynamic> toJson() => _$QuizOptionToJson(this);
}

@JsonSerializable()
class QuizSubmission {
  final String quizId;
  final List<QuizAnswer> answers;

  QuizSubmission({
    required this.quizId,
    required this.answers,
  });

  factory QuizSubmission.fromJson(Map<String, dynamic> json) => _$QuizSubmissionFromJson(json);
  Map<String, dynamic> toJson() => _$QuizSubmissionToJson(this);
}

@JsonSerializable()
class QuizAnswer {
  final String questionId;
  final String optionId;

  QuizAnswer({
    required this.questionId,
    required this.optionId,
  });

  factory QuizAnswer.fromJson(Map<String, dynamic> json) => _$QuizAnswerFromJson(json);
  Map<String, dynamic> toJson() => _$QuizAnswerToJson(this);
}

@JsonSerializable()
class QuizResult {
  final String id;
  final String quizId;
  final String quizTitle;
  final int score;
  final String result;
  final String interpretation;
  final String completedAt;

  QuizResult({
    required this.id,
    required this.quizId,
    required this.quizTitle,
    required this.score,
    required this.result,
    required this.interpretation,
    required this.completedAt,
  });

  factory QuizResult.fromJson(Map<String, dynamic> json) => _$QuizResultFromJson(json);
  Map<String, dynamic> toJson() => _$QuizResultToJson(this);
}

@JsonSerializable()
class QuizListResponse {
  final List<Quiz> quizzes;
  final int total;
  final int page;
  final int pageSize;

  QuizListResponse({
    required this.quizzes,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory QuizListResponse.fromJson(Map<String, dynamic> json) => 
      _$QuizListResponseFromJson(json);
  Map<String, dynamic> toJson() => _$QuizListResponseToJson(this);
}

@JsonSerializable()
class QuizHistoryResponse {
  final List<QuizResult> results;
  final int total;
  final int page;
  final int pageSize;

  QuizHistoryResponse({
    required this.results,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory QuizHistoryResponse.fromJson(Map<String, dynamic> json) => 
      _$QuizHistoryResponseFromJson(json);
  Map<String, dynamic> toJson() => _$QuizHistoryResponseToJson(this);
}

File: backend/internal/dto/quiz.go
package dto

import (
	"github.com/go-playground/validator/v10"
)

// QuizRequest 測驗請求 (管理員用)
type QuizRequest struct {
	Title       string              `json:"title" binding:"required,max=200" validate:"required,max=200"`
	Description string              `json:"description" binding:"omitempty" validate:"omitempty"`
	Category    string              `json:"category" binding:"required,max=50" validate:"required,max=50"`
	Questions   []QuizQuestionRequest `json:"questions" binding:"required,min=1" validate:"required,min=1"`
}

// QuizQuestionRequest 測驗題目請求
type QuizQuestionRequest struct {
	Question string   `json:"question" binding:"required" validate:"required"`
	Options  []string `json:"options" binding:"required,min=2" validate:"required,min=2"`
	OrderNum int      `json:"order_num" binding:"required,min=1" validate:"required,min=1"`
}

// QuizSubmissionRequest 測驗提交請求
type QuizSubmissionRequest struct {
	QuizID  string             `json:"quiz_id" binding:"required,uuid" validate:"required,uuid"`
	Answers map[string]int     `json:"answers" binding:"required" validate:"required"` // question_id -> option_index
}

// QuizListRequest 測驗列表請求
type QuizListRequest struct {
	Category string `json:"category" form:"category" binding:"omitempty" validate:"omitempty"`
	Page     int    `json:"page" form:"page" binding:"omitempty,min=1" validate:"omitempty,min=1"`
	Limit    int    `json:"limit" form:"limit" binding:"omitempty,min=1,max=50" validate:"omitempty,min=1,max=50"`
}

// QuizResponse 測驗回應
type QuizResponse struct {
	ID          string              `json:"id"`
	Title       string              `json:"title"`
	Description string              `json:"description,omitempty"`
	Category    string              `json:"category"`
	Questions   []QuizQuestionResponse `json:"questions,omitempty"` // 列表中不包含，詳情中包含
	IsActive    bool                `json:"is_active"`
	CreatedAt   string              `json:"created_at"`
}

// QuizQuestionResponse 測驗題目回應
type QuizQuestionResponse struct {
	ID       string   `json:"id"`
	Question string   `json:"question"`
	Options  []string `json:"options"`
	OrderNum int      `json:"order_num"`
}

// QuizSubmissionResponse 測驗提交回應
type QuizSubmissionResponse struct {
	ID          string `json:"id"`
	QuizTitle   string `json:"quiz_title"`
	Score       int    `json:"score"`
	Result      string `json:"result"`
	CompletedAt string `json:"completed_at"`
	CreatedAt   string `json:"created_at"`
}

// QuizListResponse 測驗列表回應
type QuizListResponse struct {
	Quizzes    []QuizResponse `json:"quizzes"`
	Total      int64          `json:"total"`
	Page       int            `json:"page"`
	Limit      int            `json:"limit"`
	TotalPages int            `json:"total_pages"`
	HasMore    bool           `json:"has_more"`
}

// QuizHistoryResponse 測驗歷史回應
type QuizHistoryResponse struct {
	Submissions []QuizSubmissionResponse `json:"submissions"`
	Total       int64                    `json:"total"`
	Page        int                      `json:"page"`
	Limit       int                      `json:"limit"`
	TotalPages  int                      `json:"total_pages"`
	HasMore     bool                     `json:"has_more"`
}

// Validate 驗證請求資料
func (r *QuizRequest) Validate() error {
	validate := validator.New()
	return validate.Struct(r)
}

func (r *QuizSubmissionRequest) Validate() error {
	validate := validator.New()
	return validate.Struct(r)
}

func (r *QuizListRequest) Validate() error {
	validate := validator.New()
	return validate.Struct(r)
}
